python/usb_console/radio_session.py

```diff
@@ -59,27 +72,78 @@ def check_for_downlink(self):
         '''
+        
+        #connect to PAN email account  
+        mail = imaplib.IMAP4_SSL("imap.gmail.com", 993)
+        mail.login(self.username, self.password)
+        mail.select('Inbox')
 
         while self.running_logger:
-            # TODO implement
-            pass
+            #.search() searches from mail. Data gives id's of all emails.
+            type, data = mail.search(None, '(FROM "sbdservice@sbd.iridium.com")', '(UNSEEN)')
+            mail_ids = data[0]
+            id_list = mail_ids.split()
+
+            for num in id_list:
+                #.fetch() fetches the mail for given id where 'RFC822' is an Internet 
+                # Message Access Protocol.
+                typ, data = mail.fetch(num,'(RFC822)')
+                
+                #go through each component of data and print sender and subject line
+                for response_part in data:
+                    if isinstance(response_part, tuple):
+                        # converts message from byte literal to string removing b''
+                        msg = email.message_from_string(response_part[1].decode('utf-8'))
+                        #get the file attached to the message
+                        for part in msg.walk():
+                            
+                            if part.get_content_maintype() == 'multipart':
+                                continue
+
+                            if part.get('Content-Disposition') is None:
+                                continue
+                            
+                            #check if there is an attachment
+                            if part.get_filename() is not None:
+                                #get json data from email attachment
+                                attachmentContents=part.get_payload(decode=True).decode('utf8')
+                                data=json.loads(attachmentContents)
+                                
+                                #iterate through each statefield key of the dictionary
+                                logline=""
+                                for key in self.statefields:
+                                    #if the value for a statefield exists in the json object recieved, update the dictionary and print it to the logger
+                                    if data.get(key)is not None:
+                                        newvalue=data[key]
+                                        self.statefields[key]=newvalue
+                                        logline+=key+": "+str(newvalue)+"\n"
+                                self.logger.put(logline)
 
     def read_state(self, field, timeout=None):
         '''
         Read state.
         
         Read the latest value of the state field received via downlink with the name "field".
         '''
-        raise NotImplementedError
+        return self.statefields[field]
 
     def write_multiple_states(self, fields, vals, timeout=None):
```

Comment:
In the future, we want to revisit this function to add the following features:
- Wait for a confirmation email from Iridium that the Iridium satellite network received the packet. The email returns an MTMSN number, which stands for "Mobile-Terminated Message Number" number. This number should be saved somewhere as a member variable of this class because:
- If you check out some of the emails from Iridium containing downlinks, you'll see that they contain a line starting with MTMSN. This number corresponds to the latest uplink processed by Quake. If this number isn't equal to the MTMSN of the most recent uplink sent to the Quake, then this class should forbid its user from sending additional uplinks. The goal of this is to prevent more than one command-carrying uplink packet to be processed by the satellite at a time.

For the last bullet point, we'll need to add some email body-reading functionality to `check_downlink_thread` as well.

Can you start thinking about how to include these features into the logic?

Reply:
One more point; we want to store the MOMSN number that's received in the body of a downlink email as well, so that we have a record of it.